Use paging for fetching schema tables from the cluster

haaawk opened the pull request:
Those queries can sometimes return large number of rows
and non-paged can put unnecessary stress on the cluster.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java:
```diff
@@ -22,6 +22,7 @@
 import static com.datastax.driver.core.SchemaElement.TYPE;
```
Please add "Modified by" at the top, at line 16 (not needed in other file, as it already has this notice from before):

```
/*
 * Copyright (C) 2021 ScyllaDB
 *
 * Modified by ScyllaDB
 */
```

haaawk replied:
done

avelanarius reviewed (approved):
LGTM

tarzanek commented:
I verified and tested this change
all system_schema queries now use 10000 pageing limit - the big question is, if we want this configurable, Vlad thinks it would make sense

List of queries coming from driver after this change that still don't use any page (the peers_v2 select is weird, I guess we should get some support for this table on our side - new bug?) : 

SELECT * FROM system.local WHERE key="local"
SELECT * FROM system.peers_v2
SELECT * FROM system.peers
select cluster_name from system.local where key = "local"
USE "mykeyspace"

also note that 
select cluster_name from system.local where key = "local"
was called A LOT for having just a single session (scylla side was 4.4.3) - a bug?

I can share the session trace, or you can just repeat it with enabled 100% tracing and running a sample java app with e.g. select/insert with this driver

tarzanek commented:
attaching sample scylla side tracing logs, select, insert, select, delete, select is the sample app (at the bottom)
single node only
[tracing-java-driver.ods](https://github.com/scylladb/java-driver/files/7235835/tracing-java-driver.ods)

tarzanek reviewed (approved):
Unless we want to make the 10000 (which is double the default 5000 size) parametrized, I am OK.

Or even better - why don't we use the default paging limit here?

haaawk commented:
@Lorak-mmk has checked that this PR actually does not work because the queries are done before session is created. @Lorak-mmk could you share your findings here please.

Lorak-mmk commented:
When there is a lot of tables in Scylla, the results actually take more than one page. At the `ArrayBackedResultSet.java:122` we can see:
```
        // info can be null only for internal calls, but we don't page those. We assert
        // this explicitly because MultiPage implementation doesn't support info == null.
        assert r.metadata.pagingState == null || info != null;

        return r.metadata.pagingState == null
            ? new SinglePage(
                columnDefs, tokenFactory, protocolVersion, columnDefs.codecRegistry, r.data, info)
            : new MultiPage(
                columnDefs,
                tokenFactory,
                protocolVersion,
                columnDefs.codecRegistry,
                r.data,
                info,
                r.metadata.pagingState,
                session);
```
however, now we performed multi-page query in the session initialization - so the paging state is not null, but info (type: `ExecutionInfo`) is null. If asserts were enabled, it would fail. As they are not by default, the code continues and throws NPE later.

`MultiPage` does in fact use `info`, I don't think we can stop requiring it. I tried to put together some `ExecutionInfo` so that maybe it works.

```
private static ResultSetFuture queryAsync(
      Cluster cluster, String query, Connection connection, ProtocolVersion protocolVersion)
      throws ConnectionException, BusyConnectionException {
    ExecutionInfo info =
        new ExecutionInfo(
                0,
                0,
                ImmutableList.of(cluster.manager.metadata.getHost(connection.endPoint)),
                null,
                null)
            .with(null, null, null, new SimpleStatement(query), protocolVersion, null);
    DefaultResultSetFuture future =
        new DefaultResultSetFuture(
            null,
            protocolVersion,
            new Requests.Query(query, QueryProtocolOptions.DEFAULT.copy(10000), false));
    future.setDefaultExecutionInfo(info);
    connection.write(future);
    return future;
  }
```
(`setDefaultExecutionInfo`  is just a small method that I added to put this info in future, it will overwrite `info` if `pagingState` is not null, and `info` is null). When this code is run, there is another problem - `MultiPage` also needs `SessionManager`, which we don't have in `queryAsync` method, or `SchemaParser` at all - and I don't know how to get it. Someone who knows this codebase better should take a look.

haaawk committed:
SchemaParser: fetch schema tables sequentially

To minimize the impact of new client connecting to a cluster
with a large number of tables, this patch changes the way
schema tables are fetched by the driver.

The ideal solution would be to make the queries use paging but
that's not currently possible because of the driver architecture.

This is the first step to mitigate the problem. It makes the queries
run sequentially instead of all in parallel. Also schema data is no
longer fetched with single full scans. Instead we fetch keyspaces info
with a full scan and then fetch schema data for each keyspace separately
with a single partition reads. Keyspaces are fetched sequentially.

Signed-off-by: Piotr Jastrzebski <piotr@scylladb.com>

haaawk committed:
SchemaParser: use paging for fetching tables info

We can't use the regular paging so we emulate it with LIMIT.

Signed-off-by: Piotr Jastrzebski <piotr@scylladb.com>

haaawk committed:
SchemaParser: use paging for fetching views info

We can't use the regular paging so we emulate it with LIMIT.

Signed-off-by: Piotr Jastrzebski <piotr@scylladb.com>

haaawk committed:
SchemaParser: use paging for fetching columns info

We can't use the regular paging so we emulate it with LIMIT.

Signed-off-by: Piotr Jastrzebski <piotr@scylladb.com>

haaawk committed:
SchemaParser: use paging for fetching indexes info

We can't use the regular paging so we emulate it with LIMIT.

Signed-off-by: Piotr Jastrzebski <piotr@scylladb.com>

haaawk committed:
SchemaParser: use paging for fetching UDFs info

We can't use the regular paging so we emulate it with LIMIT.

Signed-off-by: Piotr Jastrzebski <piotr@scylladb.com>

haaawk committed:
SchemaParser: use paging for fetching UDAs info

We can't use the regular paging so we emulate it with LIMIT.

Signed-off-by: Piotr Jastrzebski <piotr@scylladb.com>

haaawk committed:
SchemaParser: use paging for fetching UDTs info

We can't use the regular paging so we emulate it with LIMIT.

Signed-off-by: Piotr Jastrzebski <piotr@scylladb.com>

haaawk committed:
Configuration: make it possible to turn off paging in schema queries

Signed-off-by: Piotr Jastrzebski <piotr@scylladb.com>
